### PR TITLE
Add camel support for ownable

### PR DESCRIPTION
**Pull request by andrew-fleming**
Fixes (partial) #596  

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+use starknet::testing::set_contract_address;
+use zeroable::Zeroable;
+
+// Dispatchers
+use openzeppelin::access::ownable::interface::IOwnableDispatcher;
+use openzeppelin::access::ownable::interface::IOwnableCamelDispatcher;
+use openzeppelin::access::ownable::interface::IOwnableDispatcherTrait;
+use openzeppelin::access::ownable::interface::IOwnableCamelDispatcherTrait;
+
+// Dual case Ownable
```
Comment (martriay) [inline src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo:15]:
```suggestion
// DualCase Ownable
```

```suggestion
// Dual case ownable
```

```suggestion
// Dual case
```

i like any of these better, or even without since it's not really adding much information other than slightly grouping

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+// Mocks
+use openzeppelin::tests::mocks::dual_ownable_mocks::SnakeOwnableMock;
+use openzeppelin::tests::mocks::dual_ownable_mocks::CamelOwnableMock;
+use openzeppelin::tests::mocks::dual_ownable_mocks::SnakeOwnablePanicMock;
+use openzeppelin::tests::mocks::dual_ownable_mocks::CamelOwnablePanicMock;
+use openzeppelin::tests::mocks::non_implementing_mock::NonImplementingMock;
+
+// Other
+use openzeppelin::tests::utils;
+use openzeppelin::tests::utils::PanicTrait;
```
Comment (martriay) [inline src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo:28]:
```suggestion
```

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    set_caller_address(OWNER());
+    let target = utils::deploy(SnakeOwnableMock::TEST_CLASS_HASH, calldata);
+    (DualCaseOwnable { contract_address: target }, IOwnableDispatcher { contract_address: target })
+}
+
+fn setup_camel() -> (DualCaseOwnable, IOwnableDispatcher) {
+    let mut calldata = ArrayTrait::new();
+    set_caller_address(OWNER());
+    let target = utils::deploy(CamelOwnableMock::TEST_CLASS_HASH, calldata);
+    (DualCaseOwnable { contract_address: target }, IOwnableDispatcher { contract_address: target })
```
Comment (martriay) [inline src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo:57]:
```suggestion
    (DualCaseOwnable { contract_address: target }, IOwnableCamelDispatcher { contract_address: target })
```
shouldn't this be `Camel`?

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+///
+/// snake_case target
+///
+
+#[test]
+#[available_gas(2000000)]
+fn test_dual_transfer_ownership() {
+    let (dispatcher, target) = setup_snake();
+    set_contract_address(OWNER()); // Bug with test-runner
```
Comment (martriay) [inline src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo:117]: what bug exactly? the problem in ERC721 is only with the camelCase methods since it's after the second `call_contract_syscall` attempt only

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+fn test_dual_transfer_ownership_exists_and_panics() {
+    let (dispatcher, _) = setup_ownable_panic();
+    dispatcher.transfer_ownership(NEW_OWNER());
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_dual_renounce_ownership() {
+    let (dispatcher, target) = setup_snake();
+    set_contract_address(OWNER()); // Bug with test-runner
```
Comment (martriay) [inline src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo:142]: same question as above

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+///
+
+#[ignore] // Bug with test-runner
+#[test]
+#[available_gas(2000000)]
+fn test_dual_transferOwnership() {
+    let (dispatcher, target) = setup_camel();
+    set_contract_address(OWNER()); // Bug with test-runner
+    dispatcher.transfer_ownership(NEW_OWNER());
+    assert(target.owner() == NEW_OWNER(), 'Should change ownership');
```
Comment (martriay) [inline src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo:175]: should we normalize the assert error string? the snake case analogous one is different

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Some error', 'ENTRYPOINT_FAILED', ))]
```
Comment (martriay) [inline src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo:180]: we're missing `test_dual_no_transferOwnership` and `test_dual_no_renounceOwnership`

**Review by martriay — COMMENTED** [review]
Looking good! Left a few questions regarding project organization and doubts regarding the test runner problems in snake_case tests

File: src/openzeppelin/access/ownable.cairo [hunk 8]
```diff
@@ -1,52 +1,5 @@
-#[contract]
-mod Ownable {
-    use starknet::ContractAddress;
-    use starknet::get_caller_address;
-    use zeroable::Zeroable;
+mod ownable;
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/ownable.cairo:1]: Unfortunately, I think we need the mod to export `Ownable`. Without the mod, we get an `Identifier not found` error. We could scrap the `use ownable::Ownable`, but then the import path will be `...ownable::ownable::Ownable`

File: src/openzeppelin/tests/access/accesscontrol.cairo [hunk 9]
```diff
@@ -0,0 +1 @@
+mod test_accesscontrol;
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/accesscontrol.cairo:1]: The idea was to mirror the oz directory, but I agree that it's enough to group them in `tests/access/`

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo (hunk 1 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: Fixed!

File: src/openzeppelin/utils/constants.cairo [hunk 10]
```diff
@@ -36,3 +36,17 @@ const IACCESSCONTROL_ID: u32 = 0x7965db0b_u32;
...
+//
+
+const OWNER_SELECTOR: felt252 = 0x2016836a56b71f0d02689e69e326f4f4c1b9057164ef592671cf0d37c8040c0;
+const TRANSFER_OWNERSHIP_SELECTOR: felt252 =
+    0x2a3bb1eaa05b77c4b0eeee0116a3177c6d62319dd7149ae148185d9e09de74a;
+const TRANSFEROWNERSHIP_SELECTOR: felt252 =
+    0x14a390f291e2e1f29874769efdef47ddad94d76f77ff516fad206a385e8995f;
+const RENOUNCE_OWNERSHIP_SELECTOR: felt252 =
+    0x52580a92c73f4428f1a260c5d768ef462b25955307de00f99957df119865d;
+const RENOUNCEOWNERSHIP_SELECTOR: felt252 =
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/utils/constants.cairo:51]: Soooo I think it'll look a bit irregular, but it's 1000% more readable. `camel_SCREAM` and `snake_SCREAM`?

File: src/openzeppelin/tests/mocks/dual_ownable_mocks.cairo [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    #[external]
+    fn renounceOwnership() {
+        Ownable::renounceOwnership();
+    }
+}
+
+// Although the following modules are designed to panic,
+// functions still need a valid return value. We chose:
+//
+// false for bool
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/tests/mocks/dual_ownable_mocks.cairo:57]: Whoops, fooled myself. And yeah, I was trying to be consistent with 721, but it doesn't add much here. Will delete

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo (hunk 7 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: Do we need them here? The camel methods aren't a part of the dual case trait

```
trait DualCaseOwnableTrait {
    fn owner(self: @DualCaseOwnable) -> ContractAddress;
    fn transfer_ownership(self: @DualCaseOwnable, new_owner: ContractAddress);
    fn renounce_ownership(self: @DualCaseOwnable);
}
```

Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: And the compiler will yell that the method is not found

Comment (martriay) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: oh yes sorry my bad

File: src/openzeppelin/utils/constants.cairo (hunk 10 above)
[the remark this answers is not in the document]
Comment (ericnordelo) [reply src/openzeppelin/utils/constants.cairo:51]: +1 to this

**Review by ericnordelo — COMMENTED** [review]
It looks really good Andrew. Besides the ignored tests I think is pretty much finished.

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo (hunk 4 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: Correct, but we are still using the workaround `set_contract_address` for the first call as opposed to `set_caller_address`. This test in particular is setting the ownable contract address to the owner's address

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo (hunk 5 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: Discussion here https://github.com/OpenZeppelin/cairo-contracts/pull/625#discussion_r1239953628

File: src/openzeppelin/tests/access/ownable/test_ownable.cairo [hunk 12]
```diff
@@ -63,6 +71,43 @@ fn test_transfer_ownership_from_nonowner() {
...
+fn test_transferOwnership_to_zero() {
+    setup();
+    Ownable::transferOwnership(ZERO());
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Caller is the zero address', ))]
+fn test_transferOwnership_from_zero() {
+    assert(Ownable::owner() == ZERO(), 'Should be zero with no owner');
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_ownable.cairo]: It's to establish that the owner is zero before the panicking call. I suppose the test confirms that anyway, so it might be redundant. Will remove

[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_ownable.cairo]: And done

File: src/openzeppelin/utils/constants.cairo (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/utils/constants.cairo]: Done! Another approach that might be less jarring would be to commit entirely to the casing e.g. `transfer_ownership_selector` and `transferOwnershipSelector`

File: src/openzeppelin/tests.cairo [hunk 13]
```diff
@@ -1,7 +1,5 @@
 mod access;
-//mod test_accesscontrol;
```
Comment (martriay) [inline src/openzeppelin/tests.cairo]: i fear this `mod access` might get confused by readers with the main access module instead of the test module. i'm not against the long list of `test_<module>.cairo` files. unsure that mimicking the file structure results in more clarity, but might be just me. @ericnordelo what do you think?

File: src/openzeppelin/tests/access/test_dual_ownable.cairo [hunk 14]
```diff
@@ -1,31 +1,23 @@
-use traits::Into;
 use array::ArrayTrait;
 use starknet::ContractAddress;
 use starknet::contract_address_const;
 use starknet::testing::set_caller_address;
 use starknet::testing::set_contract_address;
+use traits::Into;
```
Comment (martriay) [inline src/openzeppelin/tests/access/test_dual_ownable.cairo:6]: why is this needed now?

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo (hunk 4 above)
Comment (martriay) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: why do you say that's a bug? it makes sense to me that `call_contract_syscall` assumes to be in a contract's context when it's called and it forwards the current contract address as the caller.

File: src/openzeppelin/tests/access/test_dual_ownable.cairo (hunk 14 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/test_dual_ownable.cairo]: It's not. Fixed

File: src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo (hunk 4 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/access/ownable/test_dual_ownable.cairo]: As discussed offline, setting the contract address as the caller in this context is _not_ a bug. Comments are removed and there's a tmp reference to the issue raised in cairo for the ignored `test_dual_transferOwnership` and `test_dual_renounceOwnership`

File: src/openzeppelin/tests.cairo (hunk 13 above)
Comment (ericnordelo) [reply src/openzeppelin/tests.cairo]: Just for the record (I know we've discussed this already), I like the minimal grouping approach with just one directory of depth at most.

**Comment by andrew-fleming** [comment]
## Update

This PR also proposes to bump the cairo submodule to [v1.1.1](https://github.com/starkware-libs/cairo/releases/tag/v1.1.1). This version includes the `call_contract_syscall` bug fix as evidenced by the passing dual dispatcher camelCase tests (`test_dual_transferOwnership` and `test_dual_reounceOwnership`)

File: src/openzeppelin/utils.cairo [hunk 15]
```diff
@@ -1,10 +1,50 @@
...
+) -> SyscallResult<Span<felt252>> {
+    match call_contract_syscall(target, snake_selector, args) {
+        Result::Ok(r) => Result::Ok(r),
+        Result::Err(r) => {
+            if *r.at(0) == 'ENTRYPOINT_NOT_FOUND' {
+                return call_contract_syscall(target, camel_selector, args);
+            } else {
+                Result::Err(r)
+            }
+        }
```
Comment (ericnordelo) [inline src/openzeppelin/utils.cairo:22]:
```suggestion
        Result::Ok(ret) => Result::Ok(ret),
        Result::Err(errors) => {
            if *errors.at(0) == 'ENTRYPOINT_NOT_FOUND' {
                return call_contract_syscall(target, camel_selector, args);
            } else {
                Result::Err(errors)
            }
        }
```

To be consistent with the other PR suggestion

**Review by ericnordelo — COMMENTED** [review]
Looking very good! Left just a small readability suggestion, but besides that I think is good to go.

File: src/openzeppelin/utils/constants.cairo (hunk 10 above)
[the remark this answers is not in the document]
Comment (martriay) [reply src/openzeppelin/utils/constants.cairo]: an alternative i'm just thinking about now is to have an `utils/selectors.cairo` module with all selectors in there without the `_SELECTOR` posfix, then we can

```cairo
use selectors;

call_contract_syscall(target, selectors::transfer_ownership, ...)
```

@ericnordelo @andrew-fleming

File: src/openzeppelin/utils/constants.cairo [hunk 16]
```diff
@@ -41,12 +41,12 @@ const DEFAULT_ADMIN_ROLE: felt252 = 0;
 // Selectors
 //
 
-const OWNER_SELECTOR: felt252 = 0x2016836a56b71f0d02689e69e326f4f4c1b9057164ef592671cf0d37c8040c0;
-const TRANSFER_OWNERSHIP_SELECTOR: felt252 =
+const owner_SELECTOR: felt252 = 0x2016836a56b71f0d02689e69e326f4f4c1b9057164ef592671cf0d37c8040c0;
+const transfer_ownership_SELECTOR: felt252 =
```
Comment (martriay) [inline src/openzeppelin/utils/constants.cairo:45]: an alternative i'm just thinking about now is to have an `utils/selectors.cairo` module with all selectors in there without the `_SELECTOR` posfix, then we can

```cairo
use selectors;

call_contract_syscall(target, selectors::transfer_ownership, ...)
```
 
@ericnordelo @andrew-fleming

**Review by martriay — APPROVED** [review]
great work! left a final suggestion, otherwise good to merge :)

File: src/openzeppelin/utils/constants.cairo (hunk 16 above)
Comment (andrew-fleming) [reply src/openzeppelin/utils/constants.cairo:45]: +1 for `selectors::<method>`

Comment (ericnordelo) [reply src/openzeppelin/utils/constants.cairo:45]: +1

Comment (andrew-fleming) [reply src/openzeppelin/utils/constants.cairo]: Done

File: src/openzeppelin/access/ownable/dual_ownable.cairo [hunk 17]
```diff
@@ -26,18 +26,16 @@ trait DualCaseOwnableTrait {
...
             .at(0))
             .try_into()
             .unwrap()
     }
 
     fn transfer_ownership(self: @DualCaseOwnable, new_owner: ContractAddress) {
-        let snake_selector = constants::transfer_ownership_SELECTOR;
-        let camel_selector = constants::transferOwnership_SELECTOR;
+        let snake_selector = selectors::transfer_ownership;
+        let camel_selector = selectors::transferOwnership;
```
Comment (martriay) [inline src/openzeppelin/access/ownable/dual_ownable.cairo:38]: i don't think we need these bindings anymore, we can just use the selectors directly no?

**Review by martriay — APPROVED** [review]
amazing! good to merge, my latest suggestion is optional

File: src/openzeppelin/access/ownable/dual_ownable.cairo (hunk 17 above)
Comment (andrew-fleming) [reply src/openzeppelin/access/ownable/dual_ownable.cairo:38]: Agreed, good call!

File: src/openzeppelin/utils/selectors.cairo [hunk 18]
```diff
@@ -9,3 +9,31 @@ const transferOwnership: felt252 =
     0x14a390f291e2e1f29874769efdef47ddad94d76f77ff516fad206a385e8995f;
 const renounce_ownership: felt252 = 0x52580a92c73f4428f1a260c5d768ef462b25955307de00f99957df119865d;
 const renounceOwnership: felt252 = 0xd5d33d590e6660853069b37a2aea67c6fdaa0268626bc760350b590490feb5;
+
+//
+// ERC721 Selectors
```
Comment (ericnordelo) [inline src/openzeppelin/utils/selectors.cairo:14]:
```suggestion
// ERC721
```

Following the Ownable comment above

**Review by ericnordelo — APPROVED** [review]
Looking good to go Andrew! Left a small format suggestion, but I think is ready.

**Review by ericnordelo — APPROVED** [review]
LGTM!
